docs/madr/decisions/014-mesh-circuit-breaker-policy.md

```diff
@@ -0,0 +1,198 @@
+easier to differentiate underlying envoy features without the need to split policy into two separate ones.
+
+Additional changes to the original `CircuitBreaker` policy consists of:
+
+- flattening `detectors.*` options with only single (`.consecutive`) property
+- changing word `Errors` to `Failures`
+- adding `origin` suffix to the local failures detector
+
+It means `totalErrors.consecutive`, `gatewayErrors.consecutive` and `localErrors.consecutive`
+becomes: `consecutiveTotalFailures`, `consecutiveGatewayFailures` and `consecutiveLocalOriginFailures` respectively.
+
+### Current configuration
+
+Below is the sample `MeshCircuitBreaker` configuration
+
+```yaml
+conf:
+  interval: 1s
+  baseEjectionTime: 30s
+  maxEjectionPercent: 20
+  splitExternalAndLocalErrors: false
+  thresholds:
+    maxConnections: 2
+    maxPendingRequests: 2
+    maxRequests: 2
+    maxRetries: 2
+  detectors:
+    totalErrors:
+      consecutive: 20
+    gatewayErrors:
+      consecutive: 10
+    localErrors:
+      consecutive: 7
+    standardDeviation:
+      requestVolume: 10
+      minimumHosts: 5
+      factor: 1.9
+    failure:
+      requestVolume: 10
+      minimumHosts: 5
+      threshold: 85
+
+```
+
+### Specification
+
+#### Top level
+
+As a part of this MADR root `targetRef`'s kind can be configured as one
+of `Mesh|MeshSubset|MeshService|MeshServiceSubset`. When we finally introduce `MeshHTTPRoute` kind and if there will be
+a reasonable use case, in the next iteration `MeshGatewayRoute` and `MeshHTTPRoute` could be added as a viable kinds.
+
+```yaml
+targetRef:
+  kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
+  name: ...
+```
+
+#### From level
+
+`MeshCircuitBreaker` is outbound only policy, so only `to` should be configured.
+
+#### To level
+
+```yaml
+to:
+- targetRef:
+    kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
+    name: ...
+```
+
+#### Minimal configuration
+
+As all the connection pool related properties have default values, there is no need to specify any of these to apply the
+policy
+
+```yaml
+default:
+  connectionPool: { }
+```
+
+### Examples
+
+#### Default configuration
+
+##### `MeshCircuitBreaker`
+
+```yaml
+type: MeshCircuitBreaker
+mesh: default
+name: default-circuit-breaker
+spec:
+  targetRef:
+    kind: Mesh
+    name: default
+  to:
+  - targetRef:
+      kind: Mesh
+      name: default
+    default:
+      connectionPool:
+        maxConnections: 1024
+        maxPendingRequests: 1024
+        maxRetries: 3
+        maxRequests: 1024
+```
+
+#### Extensive configuration
+
+```yaml
+type: MeshCircuitBreaker
+mesh: default
+name: extensive-circuit-breaker
+spec:
+  targetRef:
+    kind: Mesh
+    name: default
+  to:
+  - targetRef:
+      kind: Mesh
+      name: default
+    default:
+      connectionPool:
+        maxConnections: 1024
+        maxPendingRequests: 1024
+        maxRetries: 3
+        maxRequests: 1024
+      outlierDetection:
```

Comment:
It's important to have a simple way to disable policies (when overriding from example). How do we do this here? If there's no clear way introducing a `disabled` might help.

Reply:
I thought about it for some time and the question I didn't find clear answer yet is, if this configuration is more for operators or more for our internal mechanisms.

According to the envoy docs:
> Circuit breakers are enabled by default and have modest default values, e.g. 1024 connections per cluster. To disable circuit breakers, set the [thresholds](https://www.envoyproxy.io/docs/envoy/latest/faq/load_balancing/disable_circuit_breaking#faq-disable-circuit-breaking) to the highest allowed values.
>
> ref. https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/circuit_breaking

so when `disabled` will be set to true, do we want to just not apply the explicit envoy configuration and then as far as I understand envoy implicit defaults will be set, or if it means we should apply high values as suggested in https://www.envoyproxy.io/docs/envoy/latest/faq/load_balancing/disable_circuit_breaking#faq-disable-circuit-breaking?

I think the first one is the correct answer, wdyt?

Reply:
Ok, we spoke with @jakubdyszkiewicz on a call about it, and I misunderstood the level of abstraction `disabled` would operate on. After clarification I agree with @jakubdyszkiewicz, and I updated the MADR